### PR TITLE
JAMES-2630 Rely on Reactor for connection retries

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java [hunk 1]
```diff
@@ -137,12 +137,12 @@ private ElasticSearchMailboxConfiguration getElasticSearchMailboxConfiguration(P
...
                                    ElasticSearchMailboxConfiguration mailboxConfiguration,
-                                   ElasticSearchQuotaConfiguration quotaConfiguration,
-                                   AsyncRetryExecutor executor) throws ExecutionException, InterruptedException {
+                                   ElasticSearchQuotaConfiguration quotaConfiguration) {
 
-        return RetryExecutorUtil.retryOnExceptions(executor, configuration.getMaxRetries(), configuration.getMinDelay(), NoNodeAvailableException.class)
-            .getWithRetry(context -> connectToCluster(configuration, mailboxConfiguration, quotaConfiguration))
-            .get();
+        return Mono.fromCallable(() -> connectToCluster(configuration, mailboxConfiguration, quotaConfiguration))
+            .retryBackoff(configuration.getMaxRetries(), Duration.ofMillis(configuration.getMinDelay()))
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java]: I'm not sure a backoff strategy makes sense for the connection retry

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ResilientClusterProvider.java [hunk 2]
```diff
@@ -47,17 +47,18 @@
     private final Cluster cluster;
 
     @Inject
-    private ResilientClusterProvider(ClusterConfiguration configuration, AsyncRetryExecutor executor) throws ExecutionException, InterruptedException {
-        cluster = RetryExecutorUtil.retryOnExceptions(executor, configuration.getMaxRetry(), configuration.getMinDelay(), NoHostAvailableException.class)
-                .getWithRetry(getClusterRetryCallable(configuration))
-                .get();
+    private ResilientClusterProvider(ClusterConfiguration configuration) {
+        cluster = Mono.fromCallable(getClusterRetryCallable(configuration))
+            .retryBackoff(configuration.getMaxRetry(), Duration.ofMillis(configuration.getMinDelay()))
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ResilientClusterProvider.java]: same comment for that: retry backoff is probably not a good fit

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java (hunk 1 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java]: Can you develop a bit more? AsyncRetry stuff was providing a retry isn't it?

**Comment by aduprat** [comment]
Was :green_apple: before CI upgrade

**Comment by chibenwa** [comment]
test this please

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ResilientClusterProvider.java [hunk 3]
```diff
@@ -47,17 +47,20 @@
...
     @Inject
-    private ResilientClusterProvider(ClusterConfiguration configuration, AsyncRetryExecutor executor) throws ExecutionException, InterruptedException {
-        cluster = RetryExecutorUtil.retryOnExceptions(executor, configuration.getMaxRetry(), configuration.getMinDelay(), NoHostAvailableException.class)
-                .getWithRetry(getClusterRetryCallable(configuration))
-                .get();
+    private ResilientClusterProvider(ClusterConfiguration configuration) {
+        Duration waitDelay = Duration.ofMillis(configuration.getMinDelay());
+        cluster = Mono.fromCallable(getClusterRetryCallable(configuration))
+            .retryBackoff(configuration.getMaxRetry(), waitDelay, waitDelay)
+            .doOnError(e -> LOGGER.warn("Error establishing Cassandra connection. Next retry scheduled in {} ms", waitDelay, e))
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ResilientClusterProvider.java]: doOnError should be before retryBackoff, isn't it ?

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**